test_conformance/api/negative_device.cpp

```diff
@@ -0,0 +1,520 @@
+                               "CL_INVALID_DEVICE when: \"in_device "
+                               "is not a valid device\"",
+                               TEST_FAIL);
+    }
+
+    err = clCreateSubDevices(deviceID, nullptr, 1, out_devices.data(), nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_VALUE,
+        "clCreateSubDevices should return CL_INVALID_VALUE when: \"values "
+        "specified in properties are not valid\" using a nullptr",
+        TEST_FAIL);
+
+    err = clCreateSubDevices(deviceID, properties, 0, out_devices.data(),
+                             nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_VALUE,
+        "clCreateSubDevices should return CL_INVALID_VALUE when: \"out_devices "
+        "is not NULL and num_devices is less than the number of sub-devices "
+        "created by the partition scheme\"",
+        TEST_FAIL);
+
+    if (supported_properties != SupportedPartitionSchemes::All_Schemes)
+    {
+        std::vector<cl_device_partition_property> invalid_properties =
+            get_invalid_properties(supported_properties
+                                   ^ SupportedPartitionSchemes::All_Schemes);
+        err = clCreateSubDevices(deviceID, invalid_properties.data(), 1,
+                                 out_devices.data(), nullptr);
+        test_failure_error_ret(
+            err, CL_INVALID_VALUE,
+            "clCreateSubDevices should return CL_INVALID_VALUE when: \"values "
+            "specified in properties are valid but not supported by the "
+            "device\"",
+            TEST_FAIL);
+    }
+
+    err = clCreateSubDevices(deviceID, properties, max_for_partition + 1,
+                             out_devices.data(), nullptr);
+    test_failure_error_ret(
+        err, CL_DEVICE_PARTITION_FAILED,
+        "clCreateSubDevices should return CL_DEVICE_PARTITION_FAILED when: "
+        "\"the partition name is supported by the implementation but "
+        "in_device could not be further partitioned\"",
+        TEST_FAIL);
+
+    constexpr cl_device_partition_property INVALID_PARTITION_PROPERTY =
+        -1; // Aribitrary Invalid number
+    properties[0] = INVALID_PARTITION_PROPERTY;
+    err = clCreateSubDevices(deviceID, properties, 1, out_devices.data(),
+                             nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_VALUE,
+        "clCreateSubDevices should return CL_INVALID_VALUE when: \"values "
+        "specified in properties are not valid\" using an invalid property",
+        TEST_FAIL);
+
+    if (supported_properties & SupportedPartitionSchemes::Counts)
+    {
+        properties[0] = CL_DEVICE_PARTITION_BY_COUNTS;
+        properties[1] = max_sub_devices + 1;
+        err = clCreateSubDevices(deviceID, properties, max_sub_devices + 1,
+                                 out_devices.data(), nullptr);
+        test_failure_error_ret(
+            err, CL_INVALID_DEVICE_PARTITION_COUNT,
```

Comment:
Note: our CPU implementation is returning `CL_DEVICE_PARTITION_FAILED` in this case (and the case below), not `CL_INVALID_DEVICE_PARTITION_COUNT`.  I suppose `CL_INVALID_DEVICE_PARTITION_COUNT` is more descriptive, but `CL_DEVICE_PARTITION_FAILED` seems like it is appropriate too.  Should we accept both error codes, or do we definitely want this case to return `CL_INVALID_DEVICE_PARTITION_COUNT`?